lib/datadog/appsec/contrib/rack/gateway/request.rb

```diff
@@ -41,7 +41,14 @@ def method
 
             def headers
               result = request.env.each_with_object({}) do |(k, v), h|
-                h[k.gsub(/^HTTP_/, '').downcase!.tr('_', '-')] = v if k =~ /^HTTP_/
+                # When multiple headers with the same name are present, they are concatenated with a comma
+                # https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+                # Because headers are case insensitive, HTTP_FOO and HTTP_Foo is the same, and should be merged
```

Comment:
When could this happen?

- The Rack means upcasing HTTP headers into the `HTTP_UPCASED_HEADER` variant
- Given that `request.env.each_with_object` iterates on the rack env (which is a hash) there can be no duplicates except for mixed case, which cannot happen when Rack compliant due to the above.
- The Rack env does not concatenate multiple identical headers into a single hash key, instead making use of a single one (can't recall if first or last).

It follows that the only thing that could fill in some `HTTP_MiXeD_CaSe` is some non-Rack compliant middleware or monkeypatch that  inject pseudo headers that actually don't exist on the HTTP request AND would execute before Datadog's middlewares, which are designed to be run first at the top of the Rack stack.

Note: According to the HTTP spec concatenation MAY happen but MUST only happen for headers that are lists. As is this code does so for ALL headers so it's very much incorrect, and possibly prone to injections or opportunities for a WAF bypass (e.g `Authorization` headers being concatenated in a way that make it valid for libddwaf rules but would in isolation - as seen by the application - be dangerous)

Reply:
I agree that this is an extremely rare edge case, that would be introduced by some sort of weird middleware, and I believe that we should not be responsible for a client modifying the core behaviour of Rack, especially when it is only hypothetical. And a quick benchmark shows that this is 50x slower than your first suggestion with a bit of modification (replacing gsub by delete_prefix and regex in condition by start_with), which I believe is more important for our clients.